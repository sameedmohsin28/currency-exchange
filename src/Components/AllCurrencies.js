import { useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { useState } from 'react';
import '../styles/AllCurrencies.css';

import logo from '../assets/logoForWhite.png';
import rightArrow from '../assets/rightArrow.png';

const AllCurrencies = () => {
  const [searchQuery, setSearchQuery] = useState('');

  const searchHandler = (e) => {
    setSearchQuery(e.target.value.trim().toUpperCase());
  };

  const usd = useSelector((store) => (store.currencyArray));

  const { isLoadingCurrencyArray } = useSelector((store) => (store));

  const searchedArray = usd.filter((ifSearched) => ifSearched.symbol.includes(searchQuery));

  const date = new Date();
  const yesterdayCalc = date - 1000 * 60 * 60 * 24 * 1;
  const yesterday = new Date(yesterdayCalc);
  const yesterdayMonth = `0${yesterday.getMonth() + 1}`.slice(-2);
  const yesterdayDate = `0${yesterday.getDate() + 1}`.slice(-2);
  const yesterdayWholeDate = `${yesterday.getFullYear()}-${yesterdayMonth}-${yesterdayDate}`;
  const numberOfDays = 10;
  const multipleDaysCalc = date - 1000 * 60 * 60 * 24 * numberOfDays;
  const multipleDays = new Date(multipleDaysCalc);
  const multipleDaysMonth = `0${multipleDays.getMonth() + 1}`.slice(-2);
  const multipleDaysDate = `0${multipleDays.getDate() + 1}`.slice(-2);
  const multipleDaysWholeDate = `${multipleDays.getFullYear()}-${multipleDaysMonth}-${multipleDaysDate}`;
  const prevDate = `&start_date=${multipleDaysWholeDate}&end_date=${yesterdayWholeDate}`;

  if (isLoadingCurrencyArray) {
    return (
      <>
        <div className="AllCurrencies">
          <header className="header">
            <Link to="/" className="logo-image-and-name">
              <img src={logo} alt="logo" className="logo-image" />
              <h1 className="logo-name">ForEx Rates</h1>
            </Link>
          </header>
          <section className="isLoading-state">
            <h3 className="isLoading-message">Loading latest conversion rates...</h3>
          </section>
        </div>
      </>
    );
  }

  if (!searchedArray.length && searchQuery.length) {
    return (
      <>
        <div className="AllCurrencies">
          <header className="header">
            <Link to="/" className="logo-image-and-name">
              <img src={logo} alt="logo" className="logo-image" />
              <h1 className="logo-name">ForEx Rates</h1>
            </Link>
          </header>
          <section className="content-section">
            <div className="search-div">
              <input type="text" placeholder="Search" onChange={searchHandler} className="search-bar" />
            </div>
            <div className="no-items-container">
              <h2 className="no-items">No item matches your search</h2>
            </div>
          </section>
        </div>
      </>
    );
  }

  return (
    <>
      <div className="AllCurrencies">
        <header className="header">
          <Link to="/" className="logo-image-and-name">
            <img src={logo} alt="logo" className="logo-image" />
            <h1 className="logo-name">ForEx Rates</h1>
          </Link>
        </header>
        <section className="content-section">
          <div className="search-div">
            <input type="text" placeholder="Search" onChange={searchHandler} className="search-bar" />
          </div>
          <div className="currency-link-div">
            {searchedArray.map((eachRate) => (
              <Link
                to="/converter"
                key={eachRate.symbol}
                state={{
                  rate: eachRate.rate,
                  symbol: eachRate.symbol,
                  timeSeries: `${prevDate}&symbols=${eachRate.symbol}`,
                  days: numberOfDays,
                }}
                className="currency-link"
              >
                <img src={rightArrow} alt="" className="right-arrow" />
                <div className="currency-symbol">
                  {eachRate.symbol}
                </div>
                <div className="currency-rate">
                  {'Exchange Rate: '}
                  {eachRate.rate}
                </div>
              </Link>
            ))}
          </div>
        </section>
      </div>
    </>
  );
};

export default AllCurrencies;
