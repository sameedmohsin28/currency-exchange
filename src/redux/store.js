import { configureStore } from '@reduxjs/toolkit';
import currencySlice from './currency/currencySlice';

const store = configureStore({
  reducer: currencySlice.reducer,
});

export default store;
