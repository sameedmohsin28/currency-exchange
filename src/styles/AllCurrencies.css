.logo-name,
.currency-list,
.currency-symbol,
.currency-rate {
  color: white;
}

.AllCurrencies {
  background-color: #ec4c8b;
  min-height: 100vh;
}

.header {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  padding: 0 20px;
}

.logo-image-and-name {
  display: flex;
  flex-direction: row;
  gap: 15px;
  align-items: center;
  text-decoration: none;
}

.logo-image {
  width: 60px;
  height: 60px;
}

.currency-list {
  text-decoration: none;
  margin-right: 20px;
}

.content-section {
  display: flex;
  flex-direction: column;

  /* padding: 0 50px; */
}

.search-div {
  display: flex;
  flex-direction: row;
  justify-content: center;
  padding: 20px;
}

.search-bar {
  height: 30px;
  padding-left: 10px;
  border: none;
}

.search-bar:focus {
  outline: none;
}

.currency-link-div {
  display: grid;
  grid-template-columns: 1fr 1fr;
  background-color: #cf4278;
}

.currency-link {
  color: black;
  text-decoration: none;
  height: 90px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
  padding: 10px;
}

.currency-link:nth-child(4n+1),
.currency-link:nth-child(4n+4) {
  background-color: #db4681;
}

.right-arrow {
  align-self: flex-end;
}

.currency-symbol {
  font-weight: 700;
}

.no-items-container {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.isLoading-state {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.isLoading-message {
  color: white;
}

.no-items {
  color: white;
  font-family: 'Trebuchet MS', 'Lucida Sans Unicode', 'Lucida Grande', 'Lucida Sans', Arial, sans-serif;
}
