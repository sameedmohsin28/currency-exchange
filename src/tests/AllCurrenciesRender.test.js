import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
// eslint-disable-next-line import/extensions
import store from '../redux/store';

import AllCurrencies from '../Components/AllCurrencies';

describe('AllCurencies rendering Validation', () => {
  test('AllCurencies should render with / url', () => {
    const treeSnapShot = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Routes>
            <Route path="/" element={<AllCurrencies />} />
          </Routes>
        </BrowserRouter>
      </Provider>,
    ).toJSON();
    expect(treeSnapShot).toMatchSnapshot();
  });
});
