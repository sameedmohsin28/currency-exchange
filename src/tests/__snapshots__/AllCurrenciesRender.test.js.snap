// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllCurencies rendering Validation AllCurencies should render with / url 1`] = `
<div
  className="AllCurrencies"
>
  <header
    className="header"
  >
    <a
      className="logo-image-and-name"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="logo"
        className="logo-image"
        src="logo.jpg"
      />
      <h1
        className="logo-name"
      >
        ForEx Rates
      </h1>
    </a>
  </header>
  <section
    className="content-section"
  >
    <div
      className="search-div"
    >
      <input
        className="search-bar"
        onChange={[Function]}
        placeholder="Search"
        type="text"
      />
    </div>
    <div
      className="currency-link-div"
    />
  </section>
</div>
`;
